File: app/src/main/java/com/app/techvalley/movies/adapters/RecyclerViewAdapter.java
package com.app.techvalley.movies.adapters;

import android.content.Context;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.app.techvalley.movies.models.Movie;
import com.app.techvalley.movies.OnRecyclerViewItemClickListener;
import com.app.techvalley.movies.R;
import com.app.techvalley.movies.network.URLConstants;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

/**
 * Created by KeshavAggarwal on 06/01/17.
 */

public class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder> {
    private ArrayList<Movie> mMovies;
    Context mContext;
    private OnRecyclerViewItemClickListener listener;
    private int verticalPosition;

    public RecyclerViewAdapter(ArrayList<Movie> movies, Context context) {
        mMovies = movies;
        mContext = context;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.horizontal_cardview_movies, parent, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        if (mMovies != null) {
            holder.movieName.setText(mMovies.get(position).getTitle());
            Picasso.get().load(URLConstants.IMAGE_BASE_URL + mMovies.get(position).getPosterPath()).into(holder.movieThumbnailImage);
            if (mMovies.get(position).getDate().length() >= 5) {
                String date = mMovies.get(position).getDate().substring(0, 4);
                holder.movieReleaseDate.setText(date);
            }
            String rating = Double.toString(mMovies.get(position).getRating());
            holder.rating.setText(rating);
            holder.cv.setOnClickListener(v -> {
                if (listener != null)
                    listener.onRecyclerViewItemClicked(verticalPosition, position, holder.movieThumbnailImage);
            });
        }
    }


    @Override
    public int getItemCount() {
        return mMovies.size();
    }


    public void setOnItemClickListener(OnRecyclerViewItemClickListener listener, int verticalPosition) {
        this.listener = listener;
        this.verticalPosition = verticalPosition;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        CardView cv;
        ImageView movieThumbnailImage;
        TextView movieName;
        TextView movieReleaseDate;
        TextView rating;

        ViewHolder(View itemView) {
            super(itemView);
            cv = itemView.findViewById(R.id.cardView);
            movieThumbnailImage = itemView.findViewById(R.id.thumbnailImageView);
            movieName = itemView.findViewById(R.id.nameTextView);
            movieReleaseDate = itemView.findViewById(R.id.releaseDateTextView);
            rating = itemView.findViewById(R.id.ratingTextView);

        }
    }

}

File: app/src/main/java/com/app/techvalley/movies/adapters/RecyclerViewAdapter.kt
package com.app.techvalley.movies.adapters

import android.content.Context
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.app.techvalley.movies.models.Movie
import com.app.techvalley.movies.OnRecyclerViewItemClickListener
import com.app.techvalley.movies.R
import com.app.techvalley.movies.network.URLConstants
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.horizontal_cardview_movies.view.*

import java.util.ArrayList

/**
 * Created by KeshavAggarwal on 06/01/17.
 */

class RecyclerViewAdapter(private val mMovies: ArrayList<Movie>?, internal var mContext: Context) : RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder>() {
    private var listener: OnRecyclerViewItemClickListener? = null
    private var verticalPosition: Int = 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.horizontal_cardview_movies, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        if (mMovies != null) {
            holder.itemView.nameTV.text = mMovies[position].title
            Picasso.get().load(URLConstants.IMAGE_BASE_URL + mMovies[position].posterPath).into(holder.itemView.thumbnailIV)
            if (mMovies[position].date.length >= 5) {
                val date = mMovies[position].date.substring(0, 4)
                holder.itemView.releaseDateTV.text = date
            }
            val rating = java.lang.Double.toString(mMovies[position].rating)
            holder.itemView.ratingTV.text = rating
            holder.itemView.setOnClickListener { v ->
                if (listener != null)
                    listener!!.onRecyclerViewItemClicked(verticalPosition, position, holder.itemView.thumbnailIV)
            }
        }
    }


    override fun getItemCount(): Int {
        return mMovies?.size ?: 0
    }


    fun setOnItemClickListener(listener: OnRecyclerViewItemClickListener, verticalPosition: Int) {
        this.listener = listener
        this.verticalPosition = verticalPosition
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

}
